datafusion/optimizer/src/optimize_projections.rs

```diff
@@ -677,6 +676,22 @@ fn outer_columns_helper(expr: &Expr, columns: &mut HashSet<Column>) -> bool {
                     .as_ref()
                     .map_or(true, |expr| outer_columns_helper(expr, columns))
         }
+        Expr::SimilarTo(similar_to) => {
+            outer_columns_helper(&similar_to.expr, columns)
+                && outer_columns_helper(&similar_to.pattern, columns)
+        }
+        Expr::TryCast(try_cast) => outer_columns_helper(&try_cast.expr, columns),
+        Expr::GetIndexedField(index) => outer_columns_helper(&index.expr, columns),
+        Expr::Not(expr)
+        | Expr::IsNotFalse(expr)
+        | Expr::IsFalse(expr)
+        | Expr::IsTrue(expr)
+        | Expr::IsNotTrue(expr)
+        | Expr::IsUnknown(expr)
+        | Expr::IsNotUnknown(expr)
+        | Expr::IsNotNull(expr)
+        | Expr::IsNull(expr)
+        | Expr::Negative(expr) => outer_columns_helper(expr, columns),
         Expr::Column(_) | Expr::Literal(_) | Expr::Wildcard { .. } => true,
         _ => false,
```

Comment:
What would you think about removing this catchall (and and thus explicitly enumerating all Expr types so this type of bug can't happen again)?

```suggestion
```

An alternate idea would be to remove this explicit walk down the tree entirely and use https://docs.rs/datafusion/latest/datafusion/logical_expr/utils/fn.expr_to_columns.html instead?

This pushdown code may predate the more generaic tree walks

Reply:
Good idea. Already removed the catchall.

> An alternate idea would be to remove this explicit walk down the tree entirely and use https://docs.rs/datafusion/latest/datafusion/logical_expr/utils/fn.expr_to_columns.html instead?
> 

the `outer_columns_helper` function is used to collect all `outer_ref_column` that `expr_to_columns` function doesn't return these columns. Because `expr_to_columns` use `TreeNode` to walks the expr tree, but TreeNode don't walk the outer ref column 
https://github.com/apache/arrow-datafusion/blob/dd4263f843e093c807d63edf73a571b1ba2669b5/datafusion/expr/src/tree_node/expr.rs#L69-L77

Reply:
> the outer_columns_helper function is used to collect all outer_ref_column that expr_to_columns function doesn't return these columns. Because expr_to_columns use TreeNode to walks the expr tree, but TreeNode don't walk the outer_ref_column

I see -- thank you for the explanation, which makes sense. It fiddled a little with the code, and I think I have a cleanup here that uses the `TreeNode` walking to avoid having to enumerate `Expr` variants https://github.com/apache/arrow-datafusion/pull/8787